### ISSUE TITLE
Eliminate long living tokens in Terraform and Terraform Cloud

### ISSUE THREAD

**Issue by hendrikhalkow** [hashicorp/terraform#32122]
_TL;DR: Implement a SPIFFE provider for Terraform and deploy SPIFFE to Terraform Cloud to eliminate long living tokens forever._

## The problem

When using Terraform Cloud to provision some kind of cloud infrastructure, you need to create tokens that Terraform Cloud uses for authentication. These tokens are long living, which is bad from a security perspective. For example, for AWS,
you need to create an IAM user, which is what you want to [avoid](https://summitroute.com/downloads/aws_security_maturity_roadmap-Summit_Route.pdf) from a security perspective.

## The solution

While cloud providers solved this problem within their infrastructure, there has been no good cross-platform solution available yet – until now. Meet [SPIFFE](https://spiffe.io), a universal identity control plane for distributed systems. For a short introduction, [watch this video](https://youtu.be/-XGKybqTfZo).

To give you an idea on how SPIFFE could help Terraform and Terraform Cloud, look  at these two examples.

### The legacy way

Until today, we need to provide long living tokens to Terraform providers, for example for AWS, we use `AWS_SECRET_ACCESS_KEY` or the `secret_key` attribute of the provider:

```hcl
provider "aws" {
  region     = "eu-central-1"
  access_key = "..."
  secret_key = "..." # long living token that we want to avoid
}
```

### The SPIFFE way

If we had SPIFFE, we could eliminate the long living token in favour of a [verifiable identity](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/), encoded into a JWT or X.509 certificate. Our Terraform Cloud run would get an identity URI, the SVID, which could look like this: `spiffe://terraform.io/my-organization/my-workspace`

As you can see in the example below, there are no more long-living tokens involved and the need to create an IAM user for Terraform Cloud would be eliminated. Instead, a short-living credential would be created by a Terraform provider for SPIFFE, which is what this feature request is about.

## What needs to be done

The Terraform Code would look like this:

```hcl
# This is what we need :)
data "spiffe_token" "this" {}

provider "aws" {
  region       = "eu-central-1"
  assume_role_with_web_identity {
    duration           = "1h" # Lifetime of short-living token
    role_arn           = "arn:aws:iam::012345678901:role/terraform"
    web_identity_token = data.spiffe_token.this.jwt
  }
}
```

On the AWS side, you would have to create an IAM role with where you allow Terraform Cloud to assume the role. You see that the subject equals our SVID, which would authorise our Terraform Cloud workspace to use this identity provider. This makes it easy to _deploy individual AWS roles for each Terraform Cloud workspace_ to strengthen security.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws:iam::012345678901:oidc-provider/oidc-discovery.terraform.io"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "oidc-discovery.terraform.io:aud": "terraform",
          "oidc-discovery.terraform.io:sub": "spiffe://terraform.io/my-organization/my-workspace"
        }
      }
    }
  ]
}
```

The rest is plain AWS: you connect this OpenID Connect provider to the IAM role `arn:aws:iam::012345678901:role/terraform` that the Terraform provider for AWS would assume, you attach an appropriate policy to the role and you are done.

### Further reading

Please have a look at the [AWS OIDC Authentication](https://spiffe.io/docs/latest/keyless/oidc-federation-aws/) example on the SPIFFE website.

## What needs to be done

We actually need three things to get rid of long-living access tokens in Terraform forever.

### Implement SPIFFE provider for Terraform

First, we need the SPIFFE provider for Terraform that allows Terraform runs to retrieve their identity. It needs to implement the `spiffe_token` data source that we used in the example above.

### Deploy SPIRE to Terraform Cloud agents

Second, we need [SPIRE](https://spiffe.io/docs/latest/spire-about/spire-concepts/) or another SPIFFE implementation deployed on Terraform Cloud agents so that the SPIFFE provider for Terraform can retrieve the identity token.
 
### Provide OpenID Connect discovery provider for Terraform Cloud
 
 Third, Terraform Cloud needs to provide an OpenID Connect discovery provider so that the counterpart, e.g. the AWS OpenID Connect provider, can retrieve all the relevant identity information.

_I also emailed the same text to the Terraform Cloud team to implement their part of this proposal._

**Comment by apparentlymart** [comment]
Hi @hendrikhalkow! Thanks for sharing this.

Our friends over in the Terraform Cloud teams have been working on a feature called Workload Identity which sounds like it covers similar ground as what you're describing here.

It was announced at the recent HashiConf but isn't yet fully launched so there aren't any updated docs to link to yet, but the team told me that [Ned in the Cloud's Overview Video](https://www.youtube.com/watch?v=M5hJwl2ewTk&t=174s) is a good place to get an idea of its scope and setup details.

I'm not personally yet familiar with the exact details of how the Terraform Cloud feature works or how SPIFFE works, so I'm not sure how much overlap there is here. Would you consider the Terraform Cloud feature described in that video to be essentially equivalent to what you've described here, or are there some additional parts that you'd like to see?

Thanks again!

**Comment by hendrikhalkow** [comment]
Hi @apparentlymart , thank you for sharing the video. This is exactly what I was asking for. Even better because according to the example, you can use different tokens and therefore different roles for the plan and run phase.

The only missing thing in the video is a way to retrieve the workload identity token that is more terraformy than fetching it from an environment variable, e.g. a `tfe_workload_identity_token` data source, but I'm sure the team is already on this :)

I'm curious how the Terraform Cloud team implements Workload Identity. Could be that is even built on SPIFFE and SPIRE.

**Comment by crw** [comment]
Hi @hendrikhalkow, thanks for this write-up! I am working to get it in front of the correct product manager at HashiCorp responsible for that feature in Terraform Cloud. Since this is not something the core team can address directly, I am going to close this issue - not to indicate that it is done or will not be done, but simply because this team cannot move it forward and the internal Cloud team does not use this GitHub repository to track issues. Thanks again for this thorough feature request!

**Comment by hendrikhalkow** [comment]
According to the GitHub issue mentioned, you are actually almost done implementing this. You call it Workflow Identity for Terraform Cloud. Am I correct?
